roles/openshift_health_checker/openshift_checks/efk_stack.py

```diff
@@ -0,0 +1,604 @@
+        return super(EFKStack, cls).is_active(task_vars) and is_active
+
+    def run(self, tmp, task_vars):
+        '''Check various things and gather errors. Returns: result as hash'''
+        self.logging_namespace = get_var(task_vars, "openshift_logging_namespace", default="logging")
+        pods_for, error = self._get_pods(task_vars)
+        if error:
+            return {"failed": True, "changed": False, "msg": error}
+        check_error = dict(
+            elasticsearch=self.check_elasticsearch(pods_for['es'], task_vars),
+            fluentd=self.check_fluentd(pods_for['fluentd'], task_vars),
+            curator=self.check_curator(pods_for['curator']),
+            kibana=self.check_kibana(pods_for['kibana']),
+        )
+        if not check_error['kibana']:
+            check_error['kibana_route'] = self._check_kibana_route(task_vars)
+
+        errors = [err for err in check_error.values() if err]
+        if errors:
+            errors.insert(0, "The following logging stack issues were found:")
+            return {"failed": True, "changed": False, "msg": "\n-------\n".join(errors)}
+
+        # TODO(lmeyer): run it all again for the ops cluster
+        return {"failed": False, "changed": False, "msg": 'No problems found with logging stack.'}
+
+    def _get_pods(self, task_vars):
+        '''Get all pods and sort them by EFK component. Returns: pods by component, error string'''
+        pod_output = self._exec_oc("get pods -o json", [], task_vars)
+        try:
+            pods = json.loads(pod_output)
+            if not pods or not pods.get('items'):
+                raise ValueError()
+        except ValueError:
+            # successful run but non-parsing data generally means there were no pods in the namespace
+            return None, 'There are no pods in the {} namespace. Is logging deployed?'.format(self.logging_namespace)
+
+        logging_components = ['es', 'fluentd', 'kibana', 'curator']
+        pods_for = {comp: [] for comp in logging_components}
+
+        for pod in pods['items']:
+            component = pod['metadata']['labels'].get('component')
+            if component in logging_components:
+                pods_for[component].append(pod)
+
+        return pods_for, None
+
+    @staticmethod
+    def _not_running_pods(pods):
+        '''Returns: list of pods not in a ready and running state'''
+        return [
+            pod for pod in pods
+            if any(
+                container['ready'] is False
+                for container in pod['status']['containerStatuses']
+            ) or not any(
+                condition['type'] == 'Ready' and condition['status'] == 'True'
+                for condition in pod['status']['conditions']
+            )
+        ]
+
+    def _running_es_pods(self, es_pods):
+        '''Returns: list of running pods, list of errors about non-running pods'''
+        not_running = self._not_running_pods(es_pods)
+        if not_running:
```

Comment:
I guess the method could as easily be _running_pods and the logic reversed. I guess I focus too much on the negative.

The method returns one or zero errors. The calling logic is simpler if it doesn't have to reason about that. All the other methods called in `check_elasticsearch` also return a list of errors so it's easier to just roll them all together. I admit it looks a little weird on this method, but it seemed to make sense in context...